### PR TITLE
Use the Generic host as a standard hosting mechanism

### PR DESCRIPTION
**Pull request by danielmarbach**
- Uses the generic host as a standard hosting mechanism
- Drops support on docker template for NETCORE 2.1 and uses 3.1 instead
- Drops support for NET 452 and NET46

### CODE REVIEW

**Comment by danielmarbach** [comment]
So far I've smoke tested:

- Docker template on Windows containers (stop properly)
- Docker template on Linux containers (stop properly)
- Windows service template (works)
- ScAdapter service template (works)

**Comment by danielmarbach** [comment]
Should we consider dropping all net46x targets?

**Comment by bording** [comment]
@danielmarbach I looked through this and made some tweaks. The primary thing I suspect you might be wondering about is going back to the regular SDK instead of the Worker SDK. I looked through what that SDK brings in, and I also looked at the "Worker service" template in VS, and I don't think we need/want any of things that the worker SDK does.

> Should we consider dropping all net46x targets?

I think it's fine to keep those around for now since we do support them for NSB 7 still.

**Comment by bording** [comment]
While looking at this, I noticed the way that Directory.Build.props was introduced into this repo seems a bit off with respect to LangVersion.

At a global level, we're setting LangVersion to 7.3, but when the templates are used, there won't be anything in the resulting project files to control it in any way.

At the moment it looks like we're only using 7.0 features, but if we did introduce something from 7.3, and then someone tried to use the template in VS 2017, they'd end up with an error.

With the change currently in this PR, the docker template requires VS 2019, but the others should still work in VS 2017.

I don't want to add `LangVersion` to the actual template output, but maybe should set it to 7.0 in Directory.Build.props?

File: src/Templates/ScAdapterService/ScAdapterService.csproj [hunk 1]
```diff
@@ -1,16 +1,16 @@
...
 
   <PropertyGroup>
     <OutputType>Exe</OutputType>
     <TargetFramework>net462</TargetFramework>
     <TreatWarningsAsErrors>true</TreatWarningsAsErrors>
   </PropertyGroup>
-  
+
   <ItemGroup>
+    <PackageReference Include="Microsoft.Extensions.Hosting.WindowsServices" Version="3.1.2" />
```
Comment (danielmarbach) [inline src/Templates/ScAdapterService/ScAdapterService.csproj]: Can you explain why this change matters?

File: src/Tests/ApprovalFiles/TemplateTests.ScAdapterServiceDiffFramework.approved.txt [hunk 2]
```diff
@@ -11,14 +11,11 @@ namespace ScAdapterServiceDiffFramework
 {
     class AdapterHostedService : IHostedService
     {
-        // TODO: give the adapter an appropriate name
```
Comment (danielmarbach) [inline src/Tests/ApprovalFiles/TemplateTests.ScAdapterServiceDiffFramework.approved.txt]: Should we at least bring back the TODOs into the configs everywhere? Apparently previous people working on this code base felt it is important to highlight with a TODO.

File: src/Templates/NServiceBusDockerContainer/NServiceBusDockerContainer.csproj [hunk 3]
```diff
@@ -3,7 +3,6 @@
   <PropertyGroup>
     <OutputType>Exe</OutputType>
     <TargetFramework>netcoreapp3.1</TargetFramework>
-    <TreatWarningsAsErrors>true</TreatWarningsAsErrors>
```
Comment (danielmarbach) [inline src/Templates/NServiceBusDockerContainer/NServiceBusDockerContainer.csproj]: So if I understood correctly by moving this to the directory.build.props we still make sure our code has no warnings and we do not enforce warnings as errors on users using the template. So essentially the template is now slightly less opinionated?

**Comment by danielmarbach** [comment]
Regarding the worker SDK. I oriented myself based on the official guidance for background tasks

https://docs.microsoft.com/en-us/aspnet/core/fundamentals/host/hosted-services?view=aspnetcore-3.1&tabs=visual-studio#worker-service-template

https://docs.microsoft.com/en-us/aspnet/core/host-and-deploy/windows-service?view=aspnetcore-3.1&tabs=visual-studio#worker-service-template

https://devblogs.microsoft.com/aspnet/net-core-workers-as-windows-services/

so if you are suggestion we should not use it I think it would be great to elaborate what you mean with

> I looked through what that SDK brings in, and I also looked at the "Worker service" template in VS, and I don't think we need/want any of things that the worker SDK does.

because we are essentially deviating from the guidance and that should be done with a reasons that I would like to understand. Thanks for shedding light on this

File: src/Templates/ScAdapterService/ScAdapterService.csproj (hunk 1 above)
Comment (danielmarbach) [reply src/Templates/ScAdapterService/ScAdapterService.csproj]: Reason: Alphabetic sorting

File: src/Tests/ApprovalFiles/TemplateTests.ScAdapterServiceDiffFramework.approved.txt (hunk 2 above)
Comment (danielmarbach) [reply src/Tests/ApprovalFiles/TemplateTests.ScAdapterServiceDiffFramework.approved.txt]: Reason: This class is already not really annotated with TODOs so having one less is not a problem

File: src/Templates/NServiceBusDockerContainer/NServiceBusDockerContainer.csproj (hunk 3 above)
Comment (danielmarbach) [reply src/Templates/NServiceBusDockerContainer/NServiceBusDockerContainer.csproj]: So we discussed and we think the goal of this template is not to have an opinionated stance on treat warning as errors therefore we'll remove it

**Comment by danielmarbach** [comment]
We decided to stay for now with the standard SDK and can always use the worker template if we can really benefit from it later
